From 15fc641e728eda263e47479b74f0569467d9a685 Mon Sep 17 00:00:00 2001
From: Nathan Muggli <nmuggli@google.com>
Date: Tue, 16 Sep 2025 14:40:44 -0600
Subject: [PATCH] print: Prevent cross-user icon access

A specially crafted URI could cause Uri.getEncodedUserInfo to return an
incorrect value.  Instead, use the more robust getUserIdFromAuthority.

Additionally, refactored this check slightly to address a few previous
CL comments.

Bug: 376462130
Test: Manual test using steps from bug
Flag: EXEMPT BUGFIX
Change-Id: I020c64c8dd1b755877826c7fc7dadecae98dd5f4
(cherry picked from commit 8d40513846ee77a70f17e392ef9a9fc4ac254e44)
---
 .../server/print/PrintManagerService.java     | 24 +++++++------------
 1 file changed, 9 insertions(+), 15 deletions(-)

diff --git a/services/print/java/com/android/server/print/PrintManagerService.java b/services/print/java/com/android/server/print/PrintManagerService.java
index 4a8d73d23904..65dd7930df72 100644
--- a/services/print/java/com/android/server/print/PrintManagerService.java
+++ b/services/print/java/com/android/server/print/PrintManagerService.java
@@ -27,6 +27,7 @@ import android.annotation.UserIdInt;
 import android.app.ActivityManager;
 import android.app.admin.DevicePolicyManagerInternal;
 import android.content.ComponentName;
+import android.content.ContentProvider;
 import android.content.Context;
 import android.content.Intent;
 import android.content.pm.PackageManager;
@@ -255,7 +256,7 @@ public final class PrintManagerService extends SystemService {
             final long identity = Binder.clearCallingIdentity();
             try {
                 Icon icon = userState.getCustomPrinterIcon(printerId);
-                return validateIconUserBoundary(icon);
+                return validateIconUserBoundary(icon, resolvedUserId);
             } finally {
                 Binder.restoreCallingIdentity(identity);
             }
@@ -268,25 +269,18 @@ public final class PrintManagerService extends SystemService {
          * @param icon
          * @return icon (validated)
          */
-        private Icon validateIconUserBoundary(Icon icon) {
+        private Icon validateIconUserBoundary(Icon icon, int resolvedCallingId) {
             // Refer to Icon#getUriString for context. The URI string is invalid for icons of
             // incompatible types.
             if (icon != null && (icon.getType() == Icon.TYPE_URI
                     || icon.getType() == Icon.TYPE_URI_ADAPTIVE_BITMAP)) {
-                String encodedUser = icon.getUri().getEncodedUserInfo();
 
-                // If there is no encoded user, the URI is calling into the calling user space
-                if (encodedUser != null) {
-                    int userId = Integer.parseInt(encodedUser);
-                    // resolve encoded user
-                    final int resolvedUserId = resolveCallingUserEnforcingPermissions(userId);
-
-                    synchronized (mLock) {
-                        // Only the current group members can get the printer icons.
-                        if (resolveCallingProfileParentLocked(resolvedUserId)
-                                != getCurrentUserId()) {
-                            return null;
-                        }
+                final int iconUserId = ContentProvider.getUserIdFromAuthority(
+                        icon.getUri().getAuthority(), resolvedCallingId);
+                synchronized (mLock) {
+                    // Only the current group members can get the printer icons.
+                    if (resolveCallingProfileParentLocked(iconUserId) != getCurrentUserId()) {
+                        return null;
                     }
                 }
             }
-- 
2.52.0.rc1.455.g30608eb744-goog

